### ISSUE TITLE
Have a single base scenario, CLI to define a settlement

### ISSUE THREAD

**Issue by shirishag75** [mars-sim/mars-sim#260]
@mokun I guess we have talked about it earlier. More and more I think, I find the single  base scenario an interesting idea and something which can help us work out chinks in the system.  Let me elaborate the idea a bit more. @Urwumpe and @larsnaesbye maybe you both can chime in as well. 

a. The single base scenario is when you are arriving first to mars and have a single base. with either some pre-fab structures or robots who have been doing some things to have a base operational. 

b. So you and your team arrive (say a tem of 6-7 odd people) with the idea of making a permanent base . So here the objective is known. 

c. The first order of business is having enough beds, a kitchen area, a hydroponics garden, and regolith which is processed to make more structures. Also EVA suit manufacture, repair and environmental and structural controls are of the highest order. You could have repair drones/robots who could help you in carrying some of the tasks. For e.g. mining regolith and hauling tasks etc. 

d. So even if you are 6-7 people, the idea is to make quarters so 12 people can sleep at once. Similarly for kitchen, medical bay, EVA suits and things like that. 

e. The interesting part would be should people who are doing some tasks become more proficient and that's the experience or should they also be allowed to apply points in other areas as well. I would go for the latter because  almost all the people would be doing 2-3 jobs and would need to be jack of all trades rather than specialists. That perhaps will be possible when you have a larger team. 

I don't see any downside to this but only upsides, the upside at least from a console edition user, the savegame will be smaller and you will get messages from only a single base so you can find problems much more sooner rather than later. 

The other part is you could have multiple winnable scenarios for the single base, with each scenario win, a new mechanic is introduced and you have a choice of resource or people to have on mars . 

Also should our hero or heroine or the protagnist should done the mantle of commander from starting or could he be a nobody. A lowly technician or horticulturist or whatever who slowly steps up and takes command decisions ?

I know most RPG games take this route and while it may not be the reality from day 01, the captain could rely more and more on the protagnist to take correct decisions for the colony. 

All of the above should make troubleshooting a much more easier way than it currently is atm. If instead of being bombarded by messages from 6 different bases on the console, if am only following a single base, it's going to be easier for me to figure out what is going wrong in any situation rather than trying to figure out from the morasses of 6 bases all at once. 

Also use the savegame structure as has been used currently, the only difference is all the other bases and structures would have either 0 or null as their state. The other bases would open as our protagnist and his team progresses through the different scenarios. 

Let me know what do all of you think ?

**Comment by mokun** [comment]
> a. The single base scenario is when you are arriving first to mars and have a single base. with either some pre-fab structures or robots who have been doing some things to have a base operational.

We already have the settlement templates specifically designed to house a standard set of features and for hosting x number of colonists.

Create a settlement using Phase 1 (or 1A, 1B, 1C, 1D) template will get you started with a base furnished for 4 persons.

Create a settlement using Phase 2 (or 2A, 2B, 2C, 2D) template will get you started with a base furnished for 8 persons.

Select a Phase 3  (or 3A, 3B, 3C, 3D) template will be good for 12 persons.

Select a Phase 4 template will be good for 24 persons.

**Comment by mokun** [comment]
> b. So you and your team arrive (say a tem of 6-7 odd people) with the idea of making a permanent base . So here the objective is known.

Yes we can discuss more on putting together a **Scenario Mode**, where the objectives will be very specific.

As you know, in **Sandbox Mode**, it's up to the player what he/she wants to achieve.

**Comment by Urwumpe** [comment]
I like the idea, could we maybe combine  this with #152  ?

**Comment by mokun** [comment]
> d. So even if you are 6-7 people, the idea is to make quarters so 12 people can sleep at once. Similarly for kitchen, medical bay, EVA suits and things like that.

For beds/quarters availability,
A `Lander Hab` building always have 4 preconfigured beds as follows from `buildings.xml`. 
```
<!-- The building provides sleeping quarters for people. -->
			<!-- beds: the number of sleeping quarters provided. -->
			<living-accommodations beds="4">
				<activity>
					<activity-spot xloc="1.06" yloc="2.45" />
					<activity-spot xloc="-.33" yloc="2.74" />
					<activity-spot xloc="-2.74" yloc=".55" />
					<activity-spot xloc="-2.6" yloc="-1.06" />
				</activity>
			</living-accommodations>

```
On the other hand, a `Residential Hab` has 5 beds. A `Residential Quarters` has 12 beds.


For medical beds,
Each `Lander Hab` building is furnished with 1 medical bed. 
A `Medical Hab` has 4 medical beds.
A `Infirmary` building has 8 medical beds. 

and so on.

**Comment by shirishag75** [comment]
I meant do we have skills to make beds by people, buildings, making machinery etc. Is that all in place ? 

I did try the sandbox mode, I didn't see anything interesting or anything unique at least in the console mode. Maybe that's something that needs to be coded in as well ?

**Comment by mokun** [comment]
> c. The first order of business is having enough beds, a kitchen area, a hydroponics garden, and regolith which is processed to make more structures. 

Yes I agree. It's like what it takes to settle in. 

Up to v3.07, there is no *assigned* bed/quarters for each settler. Whoever can find an empty/unoccupied bed at the moment will use that bed.

Since 3.1.0, we added a bed registration step and each should have his/her own bed assigned- 

`Note: there might still be some walking bug that cause people to seeminingly fall asleep in hallway. Let me know if you see it.`

However, if there are 9 people in a settlement and there are only a max of 8 available beds, then may be we can let any empty beds be available for that person if it is not in used. 

Or, we can come up with a better and more logical way to handle this. May be we can find a way to create makeshift beds in the gym or in the dining hall, etc. since nothing happens *automagically* in the programming space :D 

This is one of those little details we need to *spell out* how to deal with it. If we can clearly lay out how to approach it, I'd say we're half way there.

**Comment by mokun** [comment]
> Also EVA suit manufacture, repair and environmental and structural controls are of the highest order.

Yes, conceivably they are very important and of the highest order. 

So for the player commander, we want to find a way for him/her to make important decision on the **priority** of each of these *operations*.

For instance, how the environmental control work right now in mars-sim are in terms of a list of BUILDING FUNCTION(s) such as LIFE_SUPPORT and RESOURCE_PROCESSING

In case of RESOURCE_PROCESSING, it controls the following 4 vital functions : 
- Sabatier RWGS Reactor
- Oxygen Generation System
- Grey Water Processor
- Black Water Processor

Please read https://github.com/mars-sim/mars-sim/wiki/Building-Function for a general idea.

As each building uniquely hosts a number of those FUNCTIONS at various level, we can get an idea of the interworking of the game mechanics of mars-sim and the design approach it has developed over the years. 

This may be a little bit of *developers speak* but it's crucial for developing a way forward plan, based on where we are currently at programmatically.

**Comment by mokun** [comment]
> You could have repair drones/robots who could help you in carrying some of the tasks. For e.g. mining regolith and hauling tasks etc.

Since v3.07, we added bots. For now, we only let bots to stay indoor. 

Yes they can carry things and do certain repetitive `Task` but they cannot join `Mission`.

At one thread or two, we were talking about doing a power modeling over charging and discharging the battery on a Robot. We haven't done it yet.

When it comes to mining the minerals on a `Mining` Mission, the settlers would need to deploy the Light Utility Vehicle (LUV) with the following attachment parts 
- pneumatic drill
- backhoe
- socket wrench
- pipe wrench

In near future, we will add in the use of more tools during the mission. 

Perhaps we can use bring a robot along but we'll need to define programmatically what a robot can perform that a typical person cannot, etc.
 
In short, there are tons of detail work on modeling we can improve to mimic the real life.

**Comment by mokun** [comment]
> I like the idea, could we maybe combine this with #152 ?

Let's make it happen !

**Comment by shirishag75** [comment]
@mokun Just saw you made some commits. Let me know when you have done fully. How to invoke it and then we will go about seeing what works and what doesn't ;)

**Comment by mokun** [comment]
Sure. Just try it. The commit message includes enough info for testing it. Can you see it?

**Comment by shirishag75** [comment]
Saw it, tried it and got #262 , please fix it or tell me where I went wrong.

**Comment by shirishag75** [comment]
@mokun I was able to launch the game with the new `-template:1A -country:India -sponsor:ISRO`but I didn't see anything difference as of yet in the console/headless mode. I again was made the commander of new pompeii and all the other settlements were therein. This was supposed to be a single settlement scenario or there is still bits that needs to be coded in ?

**Comment by mokun** [comment]
> but I didn't see anything difference as of yet in the console/headless mode. I again was made the commander of new pompeii and all the other settlements were therein. 

Let's tackle ONE THING AT A TIME. 


>This was supposed to be a single settlement scenario or there is still bits that needs to be coded in ?

Yes if you start the sim that way, there will have only one settlement in that simulation session.

**Comment by shirishag75** [comment]
> > but I didn't see anything difference as of yet in the console/headless mode. I again was made the commander of new pompeii and all the other settlements were therein.
> 
> Let's tackle ONE THING AT A TIME.
> 
> > This was supposed to be a single settlement scenario or there is still bits that needs to be coded in ?
> 
> Yes if you start the sim that way, there will have only one settlement in that simulation session.

No, that is incorrect because the moment I start the simulation I get the messages from other settlements as well, not just a single settlement :(

**Comment by shirishag75** [comment]
ping... I compiled with r5249 and used -template:1 -country:India -sponsor:ISRO and still there was no difference in the scenario. 

FWIW, I tried both the sandbox mode as well as the command-mode but do not see even a single bit of difference while running in the headless mode. All the bases and teams get activated :(

**Comment by mokun** [comment]
I run the console edition with the following arguments : `-new -8192x -template:3B -sponsor:SpaceX -country:USA`

```

 ---------------  M A R S   S I M U L A T I O N   P R O J E C T  ---------------

                        * * *   Main Menu   * * *


0. Exit 
1. Start a new Sim
2. Load a saved Sim
3. Change screen resolution

Choose an option : ​1

Starting a new sim...

 ---------------  M A R S   S I M U L A T I O N   P R O J E C T  ---------------

                        * * *   Mode Selection   * * *


0. Exit
1. Command Mode
2. Sandbox Mode

Select the Game Mode: ​2

Go to Sandbox Mode...


           * * *  Sandbox Mode - Crew and Site Selection  * * *

0. Exit
1. Open site editor
2. Enable/disable alpha crew
3. Proceed

NOTE 1: The alpha crew defined in crew.xml will be loaded by default.
NOTE 2: The console editon does not have the site editor.

Enter your choice: ​3

Starting a new simulation in Sandbox Mode...



-------------------  C O N S O L E   M E N U  -------------------

  1: Connect with MarsNet
  2: Change Autosave Timer
  3: Save the Simulation
  4: Change the Simulation Speed
  5: Set Logging Options
  6: Exit the Simulation
     
Enter your choice: ​1

            << Connection to MarsNet established >>

  ------------------------------------------------------------ 
|      Ctrl + LEFT / RIGHT : scroll through input history      |
|                UP / DOWN : scroll through choices            |
|                       UP : autocomplete your input           |
  ------------------------------------------------------------ 

'/h' : help,  '/k' : keywords,  '/p' : pause/resume,  '/q' : quit

Connected with MarsNet > ​settlement
You : What are the location informations regarding the settlements ?

Settlement Location Information :


  Settlement                  Lat & long     Elev [km]
 ------------------------------------------------------------------ 
  Nicolas Town            0.00° N 0.00° E       -1.482


Connected with MarsNet > ​
```

As you can see, only one settlement is being created, namely, `Nicolas Town`

btw, can you give me some good settlement names in Indian ?

**Comment by mokun** [comment]
Inside the `/mars-sim-headless/target/` is the Console Edition. See below 👍 

![image](https://user-images.githubusercontent.com/1168584/73475657-650fbc80-4356-11ea-9164-3b06a5760a72.png)

Inside the `/mars-sim-main/target/` is the Console Edition. See below :

![image](https://user-images.githubusercontent.com/1168584/73475798-abfdb200-4356-11ea-98ac-ce3b033bbcb5.png)

Note that in both cases, I rename the long file name to just `m.jar`

**Comment by mokun** [comment]
@bevans2000 

Can we come up with a brand new wiki page to document how to start a simulation with just one single settlement ? 

Also since we already have the basic ability to start a scenario with a couple of settlement and startup conditions, can we document that as well ?

A user in Facebook were asking the question where to find the documentation. 

It does seem like how we start a single settlement is still undocumented. 

I use Eclipse launcher to do all the time for debugging in a development environment 

But if we are to run it from the last 3.7.2 compiled jar, can you show how to do it ? 

May be in future versions, we can provide some batch files as examples on how to start a single settlement or a scenario with a couple of settlements.

**Comment by mokun** [comment]
@bevans2000 

I understand that we already have the following printout of what `switches` may be needed for assembling a new settlement. 

But it would help greatly if we have a wiki page for documenting examples. Also a batch file or two for easily configuring a new settlement.

```
usage: marsProject
 -baseurl <URL to remote content>    URL to the remote content repository
                                     (defaults to master in GitHub)
 -cleanui                            Disable loading stored UI
                                     configurations
 -crew <true|false>                  Enable or disable use of the crews
 -datadir <path to data directory>   Path to the data directory for
                                     simulation files (defaults to
                                     user.home)
 -diags <<module>,<module>.....>     Enable diagnositics modules
 -help                               Display help options
 -html                               Generate HTML help
 -lat <latitude>                     Set the latitude of the new template
                                     Settlement
 -load <path to simulation file>     Load the a previously saved sim. No
                                     argument open file selection dialog.
                                     'default' will use default
 -lon <longitude>                    Set the longitude of the new template
                                     Settlement
 -new                                Enable quick start
 -noaudio                            Disable the audio
 -nogui                              Disable the main UI
 -profile                            Set up the Commander Profile
 -sandbox                            Start in Sandbox Mode
 -scenario <scenario name>           New simulation from a scenario
 -site                               Start the Scenario Editor
 -sponsor <sponsor>                  Set the sponsor for the settlement
                                     template
 -template <template name>           New simulation from a template
 -timeratio <Ratio (power of 2)>     Define the time ratio of the
                                     simulation
```

**Comment by bevans2000** [comment]
Ok makes sense. Let's create a new issue and define what page we are missing and then we can get then done.
The new packaging approach makes it much easier to start the sim now.
